### PR TITLE
Space - Charlotte 

### PR DESCRIPTION
**Pull request by FreeMonkey19**
# Assignment Submission: Hotel

Congratulations! You're submitting your assignment. Please reflect on the assignment with these questions.

## Reflection
Question | Answer
:------------- | :-------------
What was a design challenge that you encountered on this project? | Not understanding what a Date class could do for me until I was knee deep into the project. 
What was a design decision you made that changed over time over the project? | I stuck with my decisions from the beginning. But I have since changed my mind about having a Room class and not having a Date class. 
What was a concept you gained clarity on, or a learning that you'd like to share? | My Room class is not necessary. It is not doing anything that Reservation class cannot do. Room does not need to be an object inside of Reservation.new. Reservation can know its room number, cost and total cost. It makes sense to me now that a Reservation would know this. And all the repeated logic I wrote around dates needs to be in a Date class. I have a clearer understanding of Single Responsibility Principle.
What is an example of a _nominal test_ that you wrote for this assignment? What makes it a nominal case? | Calculating total cost of a reservation. It's nominal b/c its core function is just that...calculating the total cost. So the cost of a 3 night stay is $600. It's the obvious. 
What is an example of an _edge case test_ that you wrote for this assignment? What makes it an edge case? | If end_date, start_date is less than today. (?) or if end_date is less than start_date (?) I'm still not clear on edge case tests. Both of the above tests give errors.
How do you feel you did in writing pseudocode first, then writing the tests and then the code? | I think I did good with this. I used the white board a lot for this project for pseudocode. Writing tests first help me to stay focused on bite sized pieces. I definitely see the importance of this process.

### CODE REVIEW

File: lib/hotel_dispatcher.rb [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+    end 
+
+    # Wave 1 - Find all reservations for a given room and date range.
+    def find_all_res_for_room(room_num, start_date, end_date)
+      found_reservations = []
+        @reservations.each  do |reservation| 
+          next if reservation.room.room_num != room_num
+            if ((start_date > reservation.start_date && start_date < reservation.end_date) ||
+            (end_date > reservation.start_date && start_date < reservation.end_date)) || 
+            (reservation.start_date < start_date && reservation.end_date > end_date)
```
Comment (jmaddox19) [inline lib/hotel_dispatcher.rb:42]: This looks slightly different from what I see on lines 64-67 and I'm unsure why. This is a good candidate for code that would be appropriate to pull out into it's own method (probably most appropriately a method in the Reservation class).

**Comment by jmaddox19** [comment]
# Hotel






## Section 1: Major Learning Goals



| Criteria | yes/no, and optionally any details/lines of code to reference |
| --- | --- |
Practices SRP by having at least two separate classes with distinct responsibilities, and test files for these two classes | ✔️
Overall, demonstrates understanding instance variables vs. local variables. (There aren't unnecessarily too many instance variables, when it should be a local variable) | ✔️
For each test file, tests demonstrate an understanding of instantiating objects properly, and using Arrange-Act-Assert | ✔️
Practices pseudocode and TDD, and reflected on it by filling out the reflection questions | ✔️
Practices git with at least 15 small commits and meaningful commit messages | ✔️

## Section 2: Code Review and Testing Requirements

| Criteria | yes/no, and optionally any details/lines of code to reference |
| --- | --- |
There is a class that represents a reservation, and a second class that holds/manages a collection of reservations through composition (instance variable) | ✔️
The logic for checking if a reservation's date overlaps with another reservation's date is complex logic that is separated into method(s) (and potentially class(es)) | See inline comments for more context on why this isn't checked.
The logic for checking if a reservation's date overlaps with another reservation's date has unit tests | ✔️
All of the given tests run and pass | ✔️
A test coverage tool is installed and used, and shows 95% test coverage | ✔️

## Section 3: Feature Requirements

| Feature Requirement: There is a method that... | yes/no |
| --- | --- |
gives back a list of rooms, and it's tested | ✔️
creates a specific reservation for a room for a given date range, and it has nominal test cases | ✔️
creates a specific reservation for a room for a given date range, and it tests an edge case, such as no available room, or invalid date range | ✔️
gives back a list of reservations on a given date. Its tests include a test that makes several reservations for a given date | ✔️
calculates the total price for a reservation | ✔️
gives back a list of available rooms for a given date range, and it has nominal test cases | ✔️
gives back a list of available rooms for a given date range, and it has edge test cases, such as no available room, or invalid date range | 
creates a block of rooms | 
reserves a room from a block | 

## Overall Feedback

| Overall Feedback | Criteria | yes/no |
| --- | --- | --- |
| Green (Meets/Exceeds Standards) | 14+ total in all sections | ✔️
| Yellow (Approaches Standards) | 9-13 total in all sections |
| Red (Not at Standard) | 0-8 total in all sections, or assignment is breaking/doesn’t run with less than 5 minutes of debugging |

### Additional Feedback



Great work overall! You've built your first project with minimal starting code. This represents an incredible milestone in your journey, and you should be proud of yourself!

I am particularly impressed by the way that you setup multiple classes that interact with each other in understandable and appropriate ways!

I do see some room for improvement around identifying oppotunities to reuse logic by putting it in it's own method. (See my inline comment for more details.) If my comments aren't thorough enough to be clear, I'm happy to talk with you about this more. :)

## Code Style Bonus Awards



Was the code particularly impressive in code style for any of these reasons (or more...?)

| Quality | Yes? |
| --- | --- |
| Perfect Indentation | There are a couple lines that are over-indented. I use a VS Code plugin called indent-rainbow to help me identify those places.
| Elegant/Clever | 
| Descriptive/Readable | 
| Concise | 
| Logical/Organized |
